File: UVa/12086/12086.cpp
#include <bits/stdc++.h>
#define MAXN 200002

using namespace std;

int n, tree[2 * MAXN];
vector <int> a(200002);

void build(int v, int l, int r) {
    if(l == r)
        tree[v] = a[l];
    else{
        int m = l + ((r - l) >> 1);
        build(v + 1, l, m);
        build(v + 2 * (m - l + 1), m + 1, r);
        tree[v] = tree[v + 1] + tree[v + 2 * (m - l + 1)];
    }
}

int sum(int v, int L, int R, int l, int r){
    int ans;
    if(l > r)
        ans = 0;
    else if(l == L && r == R)
        ans = tree[v];
    else{
        int m = L + ((R - L) >> 1);
        ans = sum(v + 1, L, m, l, min(r, m)) + sum(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r);
    }
    return ans;
}

void update(int v, int L, int R, int pos, int val){
    if(L == R)
        tree[v] = val;
    else{
        int m = L + ((R - L) >> 1);

        if(pos <= m)
            update(v + 1, L, m, pos, val);
        else
            update(v + 2 * (m - L + 1), m + 1, R, pos, val);

        tree[v] = tree[v + 1] + tree[v + 2 * (m - L + 1)];
    }
}

int main(){
    int temp, temp2, casos = 1;
    string str;

    while(cin >> n && n){
        if(casos++ > 1)
            cout << endl;

        cout << "Case " << casos - 1 << ":" << endl;
        
        for(int i = 0; i < n; i++)
            cin >> a[i];

        build(0, 0, n - 1);
        cin >> str;

        while(str != "END"){
            cin >> temp >> temp2;
            if(str == "M")
                cout << sum(0, 0, n - 1, temp - 1, temp2 - 1) << endl;
            else
                update(0, 0, n - 1, temp - 1, temp2);
            cin >> str;
        }

    }

    return 0;
}

File: UVa/12086/12086.py
from sys import stdin

#Segment tree is represented as a list
n, MAXN = int(), 200002
tree = [0 for _ in range(MAXN * 2)]

#build the segment tree
def build(a, v, l, r):
    if l == r: tree[v] = a[l]
    else:
        m = l + ((r - l) >> 1)
        build(a, v + 1, l, m)
        build(a, v + 2 * (m - l + 1), m + 1, r)
        tree[v] = tree[v + 1] + tree[v + 2 * (m - l + 1)]

#sum query
def sum(v, L, R, l, r):
    ans = None
    if l > r: ans = 0
    elif l == L and r == R: ans = tree[v]
    else:
        m = L + ((R - L) >> 1)
        ans = sum(v + 1, L, m, l, min(r, m)) + sum(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r)
    return ans

#update query
def update(v, L, R, pos, h):
    if L == R: tree[v] = h
    else:
        m = L + ((R - L) >> 1)
        if pos <= m: update(v + 1, L, m, pos, h)
        else: update(v + 2 * (m - L + 1), m + 1, R, pos, h)
        tree[v] = tree[v + 1] + tree[v + 2 * (m - L + 1)]

n = int(input())
caso = 1

while(n != 0):
    pot = []
    i = 0

    while(i < n):
        resistencia = int(input())
        pot.append(int(resistencia))
        i += 1

    if(caso > 1) : print()
    casos = "Case {}:".format(caso)
    print(casos)
    build(pot, 0, 0, n - 1)
    linea = stdin.readline()
    linea = linea.split()
    
    while(linea[0] != "END"):
        if(linea[0] == "M"):
            print(sum(0, 0, n - 1, int(linea[1]) - 1, int(linea[2]) - 1))
        else:
            update(0, 0, n - 1, int(linea[1]) - 1, int(linea[2]))
        linea = stdin.readline()
        linea = linea.split()
    
    n = int(input())
    caso += 1